Mh 18  select healthcare network  backend

sharrisoncaktus opened the pull request:
Apart from the fact that NC and SC are hardwired into the views as "supported" states, this should meet the requirements of MH-18.

nmashton committed:
[MH-18] URL, view, HTML, and basic styles for network selection

nmashton committed:
[MH-18] HTML and styles for select and text input components

nmashton committed:
Merge remote-tracking branch 'origin/myhpom-develop' into mh-18--select-healthcare-network--frontend

nmashton committed:
[MH-18] Enforce mutual exclusion of radio buttons and options

sharrisoncaktus committed:
Merge commit '51c140258fcc54cc4604a87503a7eb2bc799587b' into mh-18--select-healthcare-network--backend

* commit '51c140258fcc54cc4604a87503a7eb2bc799587b':
  [MH-18] Enforce mutual exclusion of radio buttons and options
  [MH-18] HTML and styles for select and text input components
  [MH-18] URL, view, HTML, and basic styles for network selection

sharrisoncaktus committed:
[mh-18] remove duplication in urls, views

sharrisoncaktus committed:
[mh-18] models: HealthNetwork, which is added as a field to the UserDetails

sharrisoncaktus committed:
[mh-18]: choose_network Form, View, Template

sharrisoncaktus committed:
Merge commit 'face013755b21fcbeec7c7e72a9f92eb9ca15585' into mh-18--select-healthcare-network--backend

* commit 'face013755b21fcbeec7c7e72a9f92eb9ca15585':
  [MH-57] Add dashboard placeholder page.

# Conflicts:
#	myhpom/views/__init__.py

resolved

sharrisoncaktus committed:
[mh-18] UserDetails needs a custom_provider field for cases where the user typed their own custom provider

sharrisoncaktus committed:
Merge branch 'mh-14--create-new-account--BACKEND' into mh-18--select-healthcare-network--backend

* mh-14--create-new-account--BACKEND:
  [mh-14] renamed test_next_steps_view.py for consistency
  [mh-14] simplified validators.password_validator to give a single, complete message no matter what the error was.
  [mh-14] test_user_model
  [mh-14] documentation for myhpom.models.user. importing User from here retained
  Revert "[mh-14] "This import is ultimately just from django.contrib.auth.models import User - using that directly would probably address whatever circular import required that this import get put here, and make it clearer which model User is."-Dane"
  [mh-14] "This import is ultimately just from django.contrib.auth.models import User - using that directly would probably address whatever circular import required that this import get put here, and make it clearer which model User is."-Dane
  [mh-14] is True more pythonic less todo
  [mh-14] test_signup_view: use assertRedirects()
  [mh-14] test_post_signup_invalid: form.is_valid() is False
  [mh-14] invalid (red) form labels and tos border
  [mh-14] added test and bugfix for SignupForm.clean()
  [mh-14] State label taken from form, uppercase

# Conflicts:
#	myhpom/views/__init__.py

sharrisoncaktus committed:
Merge branch 'mh-14--create-new-account--BACKEND' into mh-18--select-healthcare-network--backend

* mh-14--create-new-account--BACKEND:
  [mh-14] missed one (state)
  [mh-14] add periods to end of SignupForm errors for consistency.
  [mh-14] restyle the (global) ul.errorlist style so that form errors display left-aligned w/o bullet
  [mh-14] SignupForm with non_field_errors() displayed
  [mh-14] signup stylesheet corrections
  [mh-14] fixed invalid box around TOS
  [mh-14] add an assertion to test_passwords_dont_match to test for the presence of a 'password_confirm' error
  [mh-14] rely on the TestCase transaction rollback, no need to delete database setup within the test case method
  [mh-14] SignupForm.clean() – simpler and more correct.

sharrisoncaktus committed:
[mh-18] add unique ('state', 'name') to HealthNetwork

sharrisoncaktus committed:
[mh-18] for choose_network, redirect to next_steps if the state is not supported

sharrisoncaktus committed:
[mh-18] updated health networks list from Slack discussion

sharrisoncaktus committed:
[mh-18] HealthNetwork validates priority in pre-save

sharrisoncaktus committed:
[mh-18] test HealthNetwork model

sharrisoncaktus committed:
[mh-18] test choose_network view

sharrisoncaktus committed:
Merge branch 'myhpom-develop' into mh-18--select-healthcare-network--backend

* myhpom-develop:
  Show that objects are ordered w/o AD templates (MH-58)
  Simplifies user reference in templates (MH-58)
  Adds logout tests (MH-58)
  Update signup tests to account for logins (MH-58)
  Adds login on signup, and logout (MH-58)
  Replaces StateAdvanceDirective with field on State model (MH-58)

# Conflicts:
#	myhpom/models/__init__.py
#	myhpom/urls.py
#	myhpom/views/__init__.py

sharrisoncaktus committed:
[mh-18] bump HealthNetwork migration to #5

sharrisoncaktus committed:
[mh-18] merge migrations

sharrisoncaktus committed:
[mh-18] choose_network now has login, so it can complete its tasks

sharrisoncaktus committed:
[mh-18] FOR THE DEMO hardwire "NC" and "SC" into the signup and choose_network views as "supported" states

sharrisoncaktus committed:
[mh-18] UserDetails.custom_provider OR UserDetails.health_network, not both.

sharrisoncaktus committed:
[mh-18] choose_network view tests

sharrisoncaktus committed:
Merge branch 'myhpom-develop' into mh-18--select-healthcare-network--backend

* myhpom-develop:
  Centers create account warning text (MH-61)
  Fixes console warning about popper.es (MH-61).

sharrisoncaktus committed:
[mh-18] final corrections to choose_network view and tests

dsummersl commented:
Can one of the admins verify this patch?

kmtracey commented on myhpom/migrations/0005_health_networks.py:
```diff
@@ -0,0 +1,72 @@
+            name='healthnetwork',
+            unique_together=set([('state', 'name')]),
+        ),
+        migrations.RunPython(load_fixture_networks, reverse_code=unload_fixture_networks),
```
So I'm learning with trying to migrate back from his territories addition that this combination of schema changes and loading data is going to be unreversible. I think the migrations here need to be broken out into the schema changes distinct from the data loading/unloading. I looks like all the shcema changes could happen in the first migration and then this RunPython get put in its own following migration?

kmtracey commented on myhpom/migrations/0006_merge.py:
```diff
@@ -0,0 +1,15 @@
+    dependencies = [
+        ('myhpom', '0004_auto_20180708_0954'),
+        ('myhpom', '0005_health_networks'),
+    ]
```
I don't believe this is a necessary migration: could just change the dependencies for the previous one. Since per the previous comment the migrations need to be split anyway if we start over by deleting these two migrations the newly-added one should have the right dependencies. The way to migrate your dev box back from these two so as to start over is to comment out the RunPytyhon so that the backwards migrate "./hstcl managepy migrate myhpom 0004" will work, then these migrations can be removed and replaced with two that first do schema alterations then load data in 2nd one.

sharrisoncaktus committed:
[MH-18] branch on existence of health networks

both signup and choose_network branch based on whether the user's state has any health networks.

kmtracey commented on myhpom/templates/myhpom/accounts/choose_network.html:
```diff
@@ -0,0 +1,135 @@
+                <!-- TODO: add real URL -->
+                <a class="btn btn-primary" href="" onClick="event.preventDefault(); document.health_network_form.submit();">
+                    Go To Dashboard
+                </a>
```
Mockup has "Submit" here, which is important to indicate the data entered will be saved I think. Also, there is supposed to be a "Skip for now" link which would allow users to move forward without doing this step...also important for demo today I think.

kmtracey committed:
[MH-18] Make migrations reversible.

sharrisoncaktus committed:
[mh-18] choose_network interface improvements:
* "Submit" instead of "Go to Dashboard"
* include "Skip for now" link.
* mutex on custom_provider <=> health_network

kmtracey commented on myhpom/views/signup.py:
```diff
@@ -24,7 +24,7 @@ def signup(request):
             login(request, user)
 
-            if user_details.state.advance_directive_template:
+            if len(user_details.state.healthnetwork_set.all()) > 0:
```
https://docs.djangoproject.com/en/1.8/ref/models/querysets/#django.db.models.query.QuerySet.exists should be used here. (not important for today but in general should always use exists vs len of the query set which can be extremely inefficient way to test for existence)

kmtracey commented on myhpom/forms/choose_network.py:
```diff
@@ -0,0 +1,8 @@
+
+class ChooseNetworkForm(forms.Form):
+    custom_provider = forms.CharField(max_length=1024, required=False)
+    health_network = forms.IntegerField(required=False)
```
Needs a clean() to ensure something has been chosen I think?

kmtracey commented on myhpom/templates/myhpom/accounts/choose_network.html:
```diff
@@ -0,0 +1,144 @@
+            </div>
+
+            <div class="modal-footer">
+                <!-- TODO: add real URL -->
```
Remove TODO at some point, since url is real now

sharrisoncaktus committed:
[mh-18] moved onClick from <a> to script

sharrisoncaktus committed:
[mh-18] protect choose_network view against null choice, and only go to next_steps on POST if a choice was made.

sharrisoncaktus committed:
[mh-18] use QuerySet.exists() rather than querying for len()
